File: Sources/Import/Services/WebImportService.swift
import Foundation
import WebKit

// MARK: - Web Import Service

/// Service for extracting article content from web pages using reader mode extraction.
actor WebImportService {

    // MARK: - Properties

    private let session: URLSession
    private let timeout: TimeInterval
    private let maxContentSize: Int64

    // JavaScript for reader mode extraction (Readability-style)
    private let readerModeJS: String = """
    (function() {
        // Simple article extraction
        function extractArticle() {
            let result = {
                title: '',
                author: '',
                siteName: '',
                publishDate: '',
                content: '',
                heroImage: ''
            };

            // Title extraction
            result.title = document.querySelector('meta[property="og:title"]')?.content
                || document.querySelector('meta[name="twitter:title"]')?.content
                || document.querySelector('h1')?.innerText
                || document.title
                || '';

            // Author extraction
            result.author = document.querySelector('meta[name="author"]')?.content
                || document.querySelector('meta[property="article:author"]')?.content
                || document.querySelector('[rel="author"]')?.innerText
                || document.querySelector('.author')?.innerText
                || document.querySelector('[class*="author"]')?.innerText
                || '';

            // Site name
            result.siteName = document.querySelector('meta[property="og:site_name"]')?.content
                || window.location.hostname.replace('www.', '')
                || '';

            // Publish date
            result.publishDate = document.querySelector('meta[property="article:published_time"]')?.content
                || document.querySelector('time')?.getAttribute('datetime')
                || document.querySelector('[class*="date"]')?.innerText
                || '';

            // Hero image
            result.heroImage = document.querySelector('meta[property="og:image"]')?.content
                || document.querySelector('meta[name="twitter:image"]')?.content
                || document.querySelector('article img')?.src
                || '';

            // Content extraction - try multiple strategies
            let article = document.querySelector('article')
                || document.querySelector('[role="main"]')
                || document.querySelector('main')
                || document.querySelector('.post-content')
                || document.querySelector('.article-content')
                || document.querySelector('.entry-content')
                || document.querySelector('.content')
                || document.querySelector('#content');

            if (article) {
                // Clone to avoid modifying original
                let clone = article.cloneNode(true);

                // Remove unwanted elements
                let removeSelectors = [
                    'script', 'style', 'nav', 'header', 'footer', 'aside',
                    '.ad', '.ads', '.advertisement', '.social-share',
                    '.comments', '.related', '.sidebar', '.navigation',
                    '[class*="share"]', '[class*="social"]', '[class*="comment"]',
                    '[class*="related"]', '[class*="recommended"]', '[class*="newsletter"]',
                    'form', 'iframe', 'video', 'audio'
                ];

                removeSelectors.forEach(selector => {
                    clone.querySelectorAll(selector).forEach(el => el.remove());
                });

                // Extract text with structure
                result.content = extractTextWithStructure(clone);
            } else {
                // Fallback: get body text
                result.content = extractTextWithStructure(document.body);
            }

            return result;
        }

        function extractTextWithStructure(element) {
            let blocks = [];
            let currentBlock = '';

            function processNode(node) {
                if (node.nodeType === Node.TEXT_NODE) {
                    let text = node.textContent.trim();
                    if (text) {
                        currentBlock += text + ' ';
                    }
                } else if (node.nodeType === Node.ELEMENT_NODE) {
                    let tag = node.tagName.toLowerCase();

                    // Skip hidden elements
                    let style = window.getComputedStyle(node);
                    if (style.display === 'none' || style.visibility === 'hidden') {
                        return;
                    }

                    // Block elements create new paragraphs
                    let blockTags = ['p', 'div', 'article', 'section', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
                                     'blockquote', 'li', 'tr', 'br', 'hr'];

                    if (blockTags.includes(tag)) {
                        if (currentBlock.trim()) {
                            blocks.push(currentBlock.trim());
                            currentBlock = '';
                        }

                        // Add heading markers
                        if (tag.match(/^h[1-6]$/)) {
                            currentBlock = '## ';
                        } else if (tag === 'blockquote') {
                            currentBlock = '> ';
                        } else if (tag === 'li') {
                            currentBlock = '• ';
                        }
                    }

                    // Process children
                    for (let child of node.childNodes) {
                        processNode(child);
                    }

                    // End block after block elements
                    if (blockTags.includes(tag) && currentBlock.trim()) {
                        blocks.push(currentBlock.trim());
                        currentBlock = '';
                    }
                }
            }

            processNode(element);

            if (currentBlock.trim()) {
                blocks.push(currentBlock.trim());
            }

            // Clean up and join
            return blocks
                .filter(b => b.length > 20 || b.startsWith('##') || b.startsWith('>'))
                .join('\\n\\n');
        }

        return JSON.stringify(extractArticle());
    })();
    """

    // MARK: - Singleton

    static let shared = WebImportService()

    // MARK: - Initialization

    init(
        timeout: TimeInterval = 30,
        maxContentSize: Int64 = 10 * 1024 * 1024 // 10 MB
    ) {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout * 2
        self.session = URLSession(configuration: config)
        self.timeout = timeout
        self.maxContentSize = maxContentSize
    }

    // MARK: - Public Methods

    /// Extract article content from a URL
    func extract(from url: URL) async throws -> ExtractionResult {
        let startTime = Date()

        // Validate URL
        guard url.scheme == "http" || url.scheme == "https" else {
            throw ExtractionError.invalidURL(url: url.absoluteString)
        }

        // Fetch HTML content
        let (data, response) = try await fetchContent(from: url)

        // Check response
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ExtractionError.networkError(underlying: "Invalid response type")
        }

        switch httpResponse.statusCode {
        case 200..<300:
            break // Success
        case 401, 403:
            throw ExtractionError.paywalledContent(domain: url.host ?? "unknown")
        case 429:
            let retryAfter = httpResponse.value(forHTTPHeaderField: "Retry-After")
                .flatMap { Double($0) }
            throw ExtractionError.rateLimited(retryAfter: retryAfter)
        case 404:
            throw ExtractionError.networkError(underlying: "Page not found (404)")
        case 500..<600:
            throw ExtractionError.serverError(
                statusCode: httpResponse.statusCode,
                message: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)
            )
        default:
            throw ExtractionError.serverError(statusCode: httpResponse.statusCode, message: nil)
        }

        // Check content size
        if data.count > maxContentSize {
            throw ExtractionError.contentTooLarge(
                sizeBytes: Int64(data.count),
                limitBytes: maxContentSize
            )
        }

        // Decode HTML
        guard let html = String(data: data, encoding: .utf8)
                ?? String(data: data, encoding: .isoLatin1) else {
            throw ExtractionError.parsingFailed(reason: "Unable to decode HTML content")
        }

        // Extract using WKWebView (on main thread)
        let extraction = try await extractUsingWebView(html: html, baseURL: url)

        let extractionDuration = Date().timeIntervalSince(startTime)

        // Parse extracted content
        let result = try parseExtraction(
            extraction,
            sourceURL: url,
            originalSize: data.count,
            duration: extractionDuration
        )

        return result
    }

    /// Extract from raw HTML
    func extract(html: String, baseURL: URL?) async throws -> ExtractionResult {
        let startTime = Date()

        let extraction = try await extractUsingWebView(html: html, baseURL: baseURL)

        let extractionDuration = Date().timeIntervalSince(startTime)

        return try parseExtraction(
            extraction,
            sourceURL: baseURL,
            originalSize: html.utf8.count,
            duration: extractionDuration
        )
    }

    // MARK: - Private Methods

    private func fetchContent(from url: URL) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.setValue("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15", forHTTPHeaderField: "User-Agent")
        request.setValue("text/html,application/xhtml+xml", forHTTPHeaderField: "Accept")

        do {
            return try await session.data(for: request)
        } catch let error as URLError {
            switch error.code {
            case .timedOut:
                throw ExtractionError.timeout(seconds: Int(timeout))
            case .notConnectedToInternet, .networkConnectionLost:
                throw ExtractionError.networkError(underlying: "No internet connection")
            case .cannotFindHost:
                throw ExtractionError.networkError(underlying: "Cannot find host: \(url.host ?? "")")
            case .cancelled:
                throw ExtractionError.cancelled
            default:
                throw ExtractionError.networkError(underlying: error.localizedDescription)
            }
        }
    }

    @MainActor
    private func extractUsingWebView(html: String, baseURL: URL?) async throws -> [String: Any] {
        return try await withCheckedThrowingContinuation { continuation in
            let webView = WKWebView(frame: .zero)

            // Load HTML
            webView.loadHTMLString(html, baseURL: baseURL)

            // Wait for load and execute extraction
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                webView.evaluateJavaScript(self.readerModeJS) { result, error in
                    if let error = error {
                        continuation.resume(throwing: ExtractionError.parsingFailed(
                            reason: error.localizedDescription
                        ))
                        return
                    }

                    guard let jsonString = result as? String,
                          let jsonData = jsonString.data(using: .utf8),
                          let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
                        continuation.resume(throwing: ExtractionError.parsingFailed(
                            reason: "Failed to parse extraction result"
                        ))
                        return
                    }

                    continuation.resume(returning: json)
                }
            }
        }
    }

    private func parseExtraction(
        _ extraction: [String: Any],
        sourceURL: URL?,
        originalSize: Int,
        duration: TimeInterval
    ) throws -> ExtractionResult {
        let content = extraction["content"] as? String ?? ""

        // Clean content
        let cleanedContent = cleanText(content)

        guard !cleanedContent.isEmpty else {
            throw ExtractionError.emptyContent
        }

        // Parse sections
        let sections = parseSections(from: cleanedContent)

        // Parse publish date
        var publishDate: Date?
        if let dateString = extraction["publishDate"] as? String, !dateString.isEmpty {
            publishDate = parseDate(dateString)
        }

        // Parse hero image URL
        var heroImageURL: URL?
        if let imageString = extraction["heroImage"] as? String, !imageString.isEmpty {
            heroImageURL = URL(string: imageString)
        }

        // Detect language
        let language = detectLanguage(cleanedContent)

        // Build warnings
        var warnings: [ExtractionWarning] = []

        if (extraction["title"] as? String)?.isEmpty ?? true {
            warnings.append(ExtractionWarning(
                type: .missingTitle,
                message: "No title found",
                suggestion: "You can add a title manually"
            ))
        }

        if (extraction["author"] as? String)?.isEmpty ?? true {
            warnings.append(ExtractionWarning(
                type: .missingAuthor,
                message: "No author found",
                suggestion: nil
            ))
        }

        let wordCount = cleanedContent.split(separator: " ").count
        if wordCount < 100 {
            warnings.append(ExtractionWarning(
                type: .lowConfidence,
                message: "Extracted content is very short (\(wordCount) words)",
                suggestion: "The page may require JavaScript or contain mostly non-text content"
            ))
        }

        let confidence: Float = min(1.0, Float(wordCount) / 500.0)

        return ExtractionResult(
            title: extraction["title"] as? String,
            author: extraction["author"] as? String,
            siteName: extraction["siteName"] as? String,
            publishDate: publishDate,
            rawText: cleanedContent,
            sections: sections,
            language: language,
            wordCount: wordCount,
            heroImageURL: heroImageURL,
            sourceURL: sourceURL,
            metadata: ExtractionMetadata(
                extractorName: "WebImportService",
                extractionDuration: duration,
                confidence: confidence,
                warnings: warnings,
                originalContentSize: originalSize,
                extractedContentSize: cleanedContent.utf8.count
            )
        )
    }

    private func parseSections(from text: String) -> [ExtractedSection] {
        let paragraphs = text.components(separatedBy: "\n\n")
        var sections: [ExtractedSection] = []
        var currentOffset = 0

        for (index, paragraph) in paragraphs.enumerated() {
            let trimmed = paragraph.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else { continue }

            // Determine section type
            let type: ArticleSectionType
            if trimmed.hasPrefix("## ") {
                type = .heading
            } else if trimmed.hasPrefix("> ") {
                type = .blockquote
            } else if trimmed.hasPrefix("• ") {
                type = .listItem
            } else if index == 0 {
                type = .title
            } else {
                type = .paragraph
            }

            // Clean section markers
            var cleanedText = trimmed
            if cleanedText.hasPrefix("## ") {
                cleanedText = String(cleanedText.dropFirst(3))
            } else if cleanedText.hasPrefix("> ") {
                cleanedText = String(cleanedText.dropFirst(2))
            } else if cleanedText.hasPrefix("• ") {
                cleanedText = String(cleanedText.dropFirst(2))
            }

            sections.append(ExtractedSection(
                index: sections.count,
                type: type,
                text: cleanedText,
                startOffset: currentOffset,
                endOffset: currentOffset + cleanedText.count,
                attributes: [:]
            ))

            currentOffset += paragraph.count + 2
        }

        return sections
    }

    private func cleanText(_ text: String) -> String {
        var cleaned = text

        // Normalize whitespace
        cleaned = cleaned.replacingOccurrences(of: "\r\n", with: "\n")
        cleaned = cleaned.replacingOccurrences(of: "\r", with: "\n")

        // Remove excessive newlines
        while cleaned.contains("\n\n\n") {
            cleaned = cleaned.replacingOccurrences(of: "\n\n\n", with: "\n\n")
        }

        // Remove excessive spaces
        while cleaned.contains("  ") {
            cleaned = cleaned.replacingOccurrences(of: "  ", with: " ")
        }

        // Trim each line
        cleaned = cleaned
            .components(separatedBy: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .joined(separator: "\n")

        return cleaned.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func parseDate(_ string: String) -> Date? {
        let formatters: [ISO8601DateFormatter] = [
            {
                let f = ISO8601DateFormatter()
                f.formatOptions = [.withInternetDateTime]
                return f
            }(),
            {
                let f = ISO8601DateFormatter()
                f.formatOptions = [.withFullDate]
                return f
            }()
        ]

        for formatter in formatters {
            if let date = formatter.date(from: string) {
                return date
            }
        }

        // Try common date formats
        let dateFormatters = ["yyyy-MM-dd", "MM/dd/yyyy", "MMMM d, yyyy"]
        for format in dateFormatters {
            let formatter = DateFormatter()
            formatter.dateFormat = format
            if let date = formatter.date(from: string) {
                return date
            }
        }

        return nil
    }

    private func detectLanguage(_ text: String) -> String {
        // Use Natural Language framework
        let sample = String(text.prefix(1000))

        // Simple heuristic detection
        let englishWords = ["the", "is", "and", "of", "to", "in", "a", "that"]
        let words = sample.lowercased().split(separator: " ").map(String.init)

        let englishCount = words.filter { englishWords.contains($0) }.count
        let englishRatio = Float(englishCount) / Float(max(1, words.count))

        if englishRatio > 0.1 {
            return "en"
        }

        return "en" // Default to English
    }
}

File: android/app/src/main/java/com/listenai/service/import_content/WebImportService.kt
package com.listenai.service.import_content

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.util.concurrent.TimeUnit

/**
 * Extracted content from a web page
 */
data class ExtractedWebContent(
    val title: String,
    val author: String?,
    val siteName: String?,
    val content: String,
    val imageUrl: String?,
    val publishDate: String?,
    val wordCount: Int,
    val language: String
)

/**
 * Server extraction response
 */
private data class ServerExtractionResponse(
    val title: String?,
    val author: String?,
    val siteName: String?,
    val publishDate: String?,
    val content: String,
    val excerpt: String?,
    val heroImage: String?,
    val wordCount: Int,
    val language: String,
    val sourceUrl: String
)

/**
 * Service for importing content from web URLs
 */
class WebImportService {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .followRedirects(true)
        .build()

    private val gson = Gson()

    private var backendUrl: String? = null
    private var authTokenProvider: (suspend () -> String)? = null

    /**
     * Configure server-side extraction (preferred method)
     */
    fun configure(backendUrl: String, authTokenProvider: suspend () -> String) {
        this.backendUrl = backendUrl
        this.authTokenProvider = authTokenProvider
    }

    /**
     * Extract content from a URL
     */
    suspend fun extractContent(url: String): ExtractedWebContent? = withContext(Dispatchers.IO) {
        // Try server-side extraction first
        backendUrl?.let { backend ->
            authTokenProvider?.let { tokenProvider ->
                try {
                    val result = extractFromServer(url, backend, tokenProvider)
                    if (result != null) return@withContext result
                } catch (e: Exception) {
                    // Fall back to local extraction
                    println("Server extraction failed, falling back to local: ${e.message}")
                }
            }
        }

        // Fall back to local extraction
        extractLocally(url)
    }

    /**
     * Extract using server-side API
     */
    private suspend fun extractFromServer(
        url: String,
        backendUrl: String,
        tokenProvider: suspend () -> String
    ): ExtractedWebContent? {
        val token = tokenProvider()
        val requestBody = gson.toJson(mapOf("url" to url))
            .toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url("$backendUrl/api/extract")
            .header("Authorization", "Bearer $token")
            .header("Content-Type", "application/json")
            .post(requestBody)
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            return null
        }

        val responseBody = response.body?.string() ?: return null
        val serverResponse = gson.fromJson(responseBody, ServerExtractionResponse::class.java)

        return ExtractedWebContent(
            title = serverResponse.title ?: "Untitled",
            author = serverResponse.author,
            siteName = serverResponse.siteName,
            content = serverResponse.content,
            imageUrl = serverResponse.heroImage,
            publishDate = serverResponse.publishDate,
            wordCount = serverResponse.wordCount,
            language = serverResponse.language
        )
    }

    /**
     * Extract content locally using Jsoup
     */
    private fun extractLocally(url: String): ExtractedWebContent? {
        try {
            val request = Request.Builder()
                .url(url)
                .header("User-Agent", "Mozilla/5.0 (Linux; Android 13) AppleWebKit/537.36")
                .build()

            val response = client.newCall(request).execute()

            if (!response.isSuccessful) {
                return null
            }

            val html = response.body?.string() ?: return null
            val document = Jsoup.parse(html, url)

            val title = extractTitle(document)
            val author = extractAuthor(document)
            val siteName = extractSiteName(document, url)
            val content = extractMainContent(document)
            val imageUrl = extractMainImage(document)
            val publishDate = extractPublishDate(document)

            if (content.isBlank()) {
                return null
            }

            val wordCount = content.split(Regex("\\s+")).size

            return ExtractedWebContent(
                title = title,
                author = author,
                siteName = siteName,
                content = content,
                imageUrl = imageUrl,
                publishDate = publishDate,
                wordCount = wordCount,
                language = "en" // Default
            )
        } catch (e: Exception) {
            return null
        }
    }

    private fun extractSiteName(document: Document, url: String): String? {
        // Try og:site_name
        val ogSiteName = document.selectFirst("meta[property=og:site_name]")?.attr("content")
        if (!ogSiteName.isNullOrBlank()) return ogSiteName

        // Extract from URL
        return try {
            java.net.URL(url).host.removePrefix("www.")
        } catch (e: Exception) {
            null
        }
    }

    private fun extractTitle(document: Document): String {
        // Try Open Graph title first
        val ogTitle = document.selectFirst("meta[property=og:title]")?.attr("content")
        if (!ogTitle.isNullOrBlank()) return ogTitle

        // Try Twitter title
        val twitterTitle = document.selectFirst("meta[name=twitter:title]")?.attr("content")
        if (!twitterTitle.isNullOrBlank()) return twitterTitle

        // Fall back to page title
        val pageTitle = document.title()
        if (pageTitle.isNotBlank()) return pageTitle

        // Try h1
        return document.selectFirst("h1")?.text() ?: "Untitled"
    }

    private fun extractAuthor(document: Document): String? {
        // Try meta author
        val metaAuthor = document.selectFirst("meta[name=author]")?.attr("content")
        if (!metaAuthor.isNullOrBlank()) return metaAuthor

        // Try article:author
        val articleAuthor = document.selectFirst("meta[property=article:author]")?.attr("content")
        if (!articleAuthor.isNullOrBlank()) return articleAuthor

        // Try common author selectors
        val authorSelectors = listOf(
            ".author-name",
            ".author",
            "[rel=author]",
            ".byline",
            ".post-author",
            "[itemprop=author]"
        )

        for (selector in authorSelectors) {
            val author = document.selectFirst(selector)?.text()
            if (!author.isNullOrBlank()) return author
        }

        return null
    }

    private fun extractMainContent(document: Document): String {
        // Remove unwanted elements
        document.select("script, style, nav, header, footer, aside, .ads, .advertisement, .comments, .related").remove()

        // Try to find article content using common selectors
        val contentSelectors = listOf(
            "article",
            "[role=main]",
            ".post-content",
            ".article-content",
            ".entry-content",
            ".content",
            "main",
            "#content"
        )

        for (selector in contentSelectors) {
            val element = document.selectFirst(selector)
            if (element != null) {
                val text = element.text()
                if (text.length > 200) {
                    return cleanExtractedText(text)
                }
            }
        }

        // Fall back to body text
        val body = document.body()
        return cleanExtractedText(body?.text() ?: "")
    }

    private fun extractMainImage(document: Document): String? {
        // Try Open Graph image
        val ogImage = document.selectFirst("meta[property=og:image]")?.attr("content")
        if (!ogImage.isNullOrBlank()) return resolveUrl(ogImage, document.baseUri())

        // Try Twitter image
        val twitterImage = document.selectFirst("meta[name=twitter:image]")?.attr("content")
        if (!twitterImage.isNullOrBlank()) return resolveUrl(twitterImage, document.baseUri())

        // Try first large image in article
        val articleImage = document.selectFirst("article img[src]")?.attr("src")
        if (!articleImage.isNullOrBlank()) return resolveUrl(articleImage, document.baseUri())

        return null
    }

    private fun extractPublishDate(document: Document): String? {
        // Try article:published_time
        val publishedTime = document.selectFirst("meta[property=article:published_time]")?.attr("content")
        if (!publishedTime.isNullOrBlank()) return publishedTime

        // Try datePublished schema
        val datePublished = document.selectFirst("[itemprop=datePublished]")?.attr("content")
        if (!datePublished.isNullOrBlank()) return datePublished

        // Try time element
        val timeElement = document.selectFirst("time[datetime]")?.attr("datetime")
        if (!timeElement.isNullOrBlank()) return timeElement

        return null
    }

    private fun cleanExtractedText(text: String): String {
        return text
            .replace(Regex("\\s+"), " ")
            .replace(Regex("\\n{3,}"), "\n\n")
            .trim()
    }

    private fun resolveUrl(url: String, baseUrl: String): String {
        return try {
            if (url.startsWith("http://") || url.startsWith("https://")) {
                url
            } else if (url.startsWith("//")) {
                "https:$url"
            } else {
                java.net.URL(java.net.URL(baseUrl), url).toString()
            }
        } catch (e: Exception) {
            url
        }
    }
}
